Improvements to eval_code

hoodmane opened the pull request:
If code ends in semicolon suppress return value (return None).
If code ends in expression return the value of the expression (this is as before).
If code ends in await expression, return the value of awaited expression.
~~If code ends in an assignment, we return the assigned lvalue.~~
~~This works with arbitrary destructuring assignments.~~

One tiny drawback of destructuring assignment is that `l[x] = 5`
gets turned into
```python
   l[x] = 5
   l[x]
```
If `l` has bad definitions for getitem and setitem, it is possible that
the second `l[x]` yields an error. This is in any case a bug in the definition
of `l`, but it makes that bug more confusing.

I also updated it to compile with `ast.PyCF_ALLOW_TOP_LEVEL_AWAIT`.
This only works with python 3.8 and up, but I think we are intending to only
support these versions anyways. For backwards compatibility we could check
if the flag exists first.

The await handling only really will do anything once we have async/await
support, see #245.

casatir commented:
Not sure about this feature:

> If code ends in an assignment, we return the assigned lvalue.

That's not the behavior you get in interpreters or notebooks (IPython, Jupyter, Iodide, Basthon). Moreover, in Python, an assignment is not an expression and has no value. E.g. You can't do `b = (a = 5)` (which is allowed in JS).

If an assignment has no value, I am not expecting getting any value from the evaluation of a code ending with an assignment.

hoodmane commented:
Okay, will remove that then.

rth commented:
Thanks! In addition could you please add tests to `src/tests/test_pyodide.py` for each of the added private function and maybe a few tests for `eval_code`? Particularly since ast manipulation is non trivial.

hoodmane commented:
@rst I added those tests. I don't think it makes a lot of sense to test `_adjust_ast_to_store_result_helper` separately but I added a handful of tests for `_adjust_ast_to_store_result`.

rth commented on src/pyodide.py:
```diff
@@ -46,7 +47,10 @@ def eval_code(code: str, ns: Dict[str, Any]) -> None:
+    If the last nonwhitespace character of code is a semicolon return `None`.
+    If the last statement is an expression, return the result of the
+    expression. If the last statement is an assignment, we return the result
+    of the assignment (including for destructuring assignments).
```
The latest version returns None, I think?

rth commented on src/pyodide.py:
```diff
@@ -55,19 +59,44 @@ def eval_code(code: str, ns: Dict[str, Any]) -> None:
+    # res =
+    eval(compile(mod, "<exec>", mode="exec", flags=COMPILE_FLAGS), ns, ns)
+    # if iscoroutine(res):
+    #     await res
```
Please remove commented code

rth reviewed (commented):
Thanks @hoodmane ! Please merge master to fix merge conflicts and add an entry to `docs/changelog.md` with a link to this PR. Otherwise LGTM (tough I'm not an expert on AST manipulations).

casatir commented:
Hi there. I have to say that I am confused with this PR. Some changes break the default Python behavior. Why do we want this? @hoodmane can you please motivate this PR?

The evaluation of `1+1;` gives `2` in Python ; this will return `None` with your modified `eval_code`. Is that right?

rth commented:
It's true that in CPython `1+1;` returns 2. In IPython REPL (and jupyter) it returns and empty string. I'm not sure what's the right thing to do so it's more a question for you @casatir and @gzuidhof since you build more complex notebooks UI on top of pyodide.

hoodmane commented:
> can you please motivate this PR?

I can try. My main motivation was to work towards asyncio support.

As soon as we have a webloop we should be using `PyCF_ALLOW_TOP_LEVEL_AWAIT` and automatically handling awaits that appear in the repl. (I think users would be much happier with `await fetch("some_url")` than `asyncio.ensure_future(fetch("some_url")`). One thing that is necessary to support the await keyword well is to return values from it if it appears at the end of the repl. Returning a value from an await expression is definitely the right option, I think that change should stay for sure.

Of course this asyncio stuff is waiting on various other PRs, but I wanted to set up `eval_code` to support it

> The evaluation of 1+1; gives 2 in Python ; this will return None with your modified eval_code. Is that right?

This is right. I think the right behavior is to allow the semicolon to suppress the output: sometimes I call a function `f()` which has side effects and very large output for the sideffects. The large output is a nuisance and I need a way to suppress it. Without this semicolon rule I have to say `f(); None`. There is no other reason to end code with a semicolon, so this notation is available for this use. I don't see why having `1+1;` be exactly the same as `1+1` is of any use because nobody would ever type `1+1;` normally.

hoodmane commented:
The main issue from my perspective is that `eval_code` can't be (usefully) monkey patched because `run_python` stores `eval_code` into a variable at startup. Otherwise, each of us can use our own `eval_code` replacement to our taste. I will open a separate issue to fix that.

casatir commented:
Thanks @hoodmane for the explanations.

>This is right. I think the right behavior is to allow the semicolon to suppress the output: sometimes I call a function f() which has side effects and very large output for the sideffects. The large output is a nuisance and I need a way to suppress it.

You mean result? You will disable REPL showing the result but side effects remains.

> Without this semicolon rule I have to say f(); None. There is no other reason to end code with a semicolon, so this notation is available for this use. I don't see why having 1+1; be exactly the same as 1+1 is of any use because nobody would ever type 1+1; normally.

Okay, you convinced me! Semicolon is a statement separator ; putting it at the end of an expression will return the null statement, `None`. Sounds reasonable.

> The main issue from my perspective is that eval_code can't be (usefully) monkey patched because run_python stores eval_code into a variable at startup. Otherwise, each of us can use our own eval_code replacement to our taste. I will open a separate issue to fix that.

+1 for this. That's definitely a way to go. Why not create `eval_code_internal` wrapped by `eval_code` that you can monkey patch? This will bypass the reference in `runpython.c`. The monkey patch will be as simple as setting `pyodide.eval_code_internal`. Obviously, it's just name convention and it could also be the other way around if you prefer to monkey patch `eval_code`. WDYT?

hoodmane commented:
I will open a separate issue describing the appropriate changes to `runpython.c`.

hoodmane commented:
@rth Assuming the CI passes after the merge, can this be accepted now that @casatir is on board?
I added `_eval_code_async` for use once we get an `EventLoop`.

rth commented:
Thanks @hoodmane and @casatir !

> Okay, you convinced me! Semicolon is a statement separator ; putting it at the end of an expression will return the null statement, None. Sounds reasonable.

OK, merging then.
